src/interfaces/interfaces.ts

```diff
@@ -54,34 +30,142 @@ namespace interfaces {
 
     export type BindingDeactivation<T> = (injectable: T) => void | Promise<void>;
 
+    export interface ValueProvider<TActivated,TValueFrom>{
+        valueFrom:TValueFrom
+        provideValue(context:Context, childRequests:Request[]):TActivated|Promise<TActivated>
+    }
+
+    export type FactoryType = keyof Pick<BindingToSyntax<unknown>,"toFactory"|"toProvider"|"toDynamicValue">;
+
+    export interface FactoryTypeValueProvider<TActivated,TValueFrom> extends ValueProvider<TActivated,TValueFrom>{
+        factoryType:FactoryType
+    }
+
+    export interface ConstantValueProvider<TActivated>
+        extends ValueProvider<TActivated,TActivated>, Clonable<ConstantValueProvider<TActivated>>
+    {
+        type:"ConstantValue";
+    }
+
+    export interface InstanceValueProvider<TActivated>
+        extends ValueProvider<TActivated,interfaces.Newable<TActivated>>, Clonable<InstanceValueProvider<TActivated>>{
+        type:"Instance";
+    }
+
+    export interface DynamicValueProvider<TActivated> extends
```

Comment:
same here. idk having strings like this is scary

Reply:
I don't understand your position.

From before.
The BindingType is a union of strings.
The BindingTypeEnum interface has been written so that it is possible for the literal to provide the same BindingType for all members
```
//interfaces
    export type BindingType = "ConstantValue" | "Constructor" | "DynamicValue" | "Factory" |
        "Function" | "Instance" | "Invalid" | "Provider";

    export interface BindingTypeEnum {
        ConstantValue: interfaces.BindingType;
        Constructor: interfaces.BindingType;
        DynamicValue: interfaces.BindingType;
        Factory: interfaces.BindingType;
        Function: interfaces.BindingType;
        Instance: interfaces.BindingType;
        Invalid: interfaces.BindingType;
        Provider: interfaces.BindingType;
    }
//literal-types
const BindingTypeEnum: interfaces.BindingTypeEnum = {
    ConstantValue: "ConstantValue",
    Constructor: "Constructor",
    DynamicValue: "DynamicValue",
    Factory: "Factory",
    Function: "Function",
    Instance: "Instance",
    Invalid: "Invalid",
    Provider: "Provider"
};

```

There was nothing preventing
```
//literal-types
const BindingTypeEnum: interfaces.BindingTypeEnum = {
    ConstantValue: "ConstantValue",
    Constructor: "ConstantValue",
    DynamicValue: "ConstantValue",
    Factory: "ConstantValue",
    Function: "ConstantValue",
    Instance: "ConstantValue",
    Invalid: "ConstantValue",
    Provider: "ConstantValue"
};
```

Now we have the concept of a ValueProvider where each of the derivations have a type that can be used for discriminated unions. (  Although it is highly unlikely, imo, we could add another one later. ).
The BindingType is a union of these types - so no change to before, it is a union of strings. 
I have typed the BindingTypeEnum interface such that the class has to follow.

```
//interfaces
    export type ValueProviderType<TActivated> =
        ConstantValueProvider<TActivated> |
        InstanceValueProvider<TActivated> |
        DynamicValueProvider<TActivated> |
        ConstructorValueProvider<TActivated> |
        FactoryValueProvider<TActivated> |
        ProviderValueProvider<TActivated> |
        NotConfiguredValueProvider

    export type BindingType = ValueProviderType<unknown>["type"];

    export interface BindingTypeEnum {
        ConstantValue: ConstantValueProvider<unknown>["type"];
        Constructor: ConstructorValueProvider<unknown>["type"];
        DynamicValue: DynamicValueProvider<unknown>["type"];
        Factory: FactoryValueProvider<unknown>["type"];
        Instance: InstanceValueProvider<unknown>["type"];
        Provider: ProviderValueProvider<unknown>["type"];
        NotConfigured: NotConfiguredValueProvider["type"];
    }
//literal-types
const BindingTypeEnum: interfaces.BindingTypeEnum = {
    ConstantValue: "ConstantValue",
    Constructor: "Constructor",
    DynamicValue: "DynamicValue",
    Factory: "Factory",
    Instance: "Instance",
    Provider: "Provider",
    NotConfigured: "NotConfigured"
};
```

Please can you explain your concern.